compute_tools/src/bin/compute_ctl.rs

```diff
@@ -275,12 +276,77 @@ fn main() -> Result<()> {
+        //
+        // Maybe we want to do this asynchronously in the future? If so, we should probably set
+        // SkipSwapon for VMs before doing that (so that we're not running 'swapoff' during or
+        // after postgres startup).
+        // We may also *not* want to do it asynchronously, so we can guarantee that the swap is
+        // actually available before postgres starts (otherwise we may have potential issues).
+        // Or maybe the startup delay isn't worth it, and if postgres OOMs it'll restart
+        // successfully once the swap is added.
+        //
+        // ---
+        //
+        // run `/neonvm/bin/resize-swap --once {size_bytes}`
+        //
+        // Passing '--once' causes resize-swap to delete itself after successful completion, which
+        // means that if compute_ctl restarts later, we won't end up calling 'swapoff' while
+        // postgres is running.
+        //
+        // NOTE: resize-swap is dumb. If present, --once MUST be the first arg.
+        const RESIZE_SWAP_BIN: &str = "/neonvm/bin/resize-swap";
+        // use a closure to make error handling easier.
+        let result = (|| -> anyhow::Result<()> {
+            let child_result = std::process::Command::new(RESIZE_SWAP_BIN)
+                .arg("--once")
+                .arg(size_bytes.to_string())
+                .spawn();
+            let mut child = match child_result {
+                Ok(child) => child,
+                Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                    warn!("ignoring \"not found\" error from resize-swap to avoid swapoff while compute is running");
+                    return Ok(()); // returns from the closure
+                }
+                Err(e) => return Err(e).context("spawn() failed"), // returns from the closure
+            };
+
+            child
+                .wait()
+                .context("wait() failed")
+                .and_then(|status| match status.success() {
+                    true => Ok(()),
+                    false => Err(anyhow!("process exited with {status}")),
+                })
+        })();
+        match result {
+            Ok(()) => {
+                let size_gib = size_bytes as f32 / (1 << 20) as f32; // just for more coherent display.
+                info!(%size_bytes, %size_gib, "resized swap");
+            }
+            Err(err) => {
+                error!("could not run `{RESIZE_SWAP_BIN} {size_bytes}: {err:#}");
+                // TODO(#7239/sharnoff): What should we do here?
+                // Should we fail compute startup if swap resizing fails?
+                //
+                // It'd be a clearer signal, making it immediately clear if something's wrong, but
+                // swap is also often just nice-to-have, so having startup fail over a nice-to-have
+                // feature makes the system more fragile. Maybe that's ultimately the best though.
```

Comment:
I'd particularly appreciate thoughts here. Skipping postgres startup seems potentially annoying to implement right, I'm partly hesitant to do it because of that.

Maybe it matters less if/when we make swap resizing asynchronous? (although in that case, we still need to decide how to handle the failure...)

Reply:
Swap primarily functions as a reliability improvement. If we fail to resize, it's possible that we end up with no swap available for the entire lifetime of the compute, which would silently impact reliability.

This _would_ be visible in the endpoint metrics [here](https://neonprod.grafana.net/d/edc7unujh4uf4d/neon-autoscaling-per-vm-metrics?orgId=1), but I don't expect we'd get alerted about it or anything.

Reply:
Ended up implementing the "fail and don't start postgres if swap resizing fails" behavior in 470c7902df48e1b632f4e59b0db61bba1d040665